### PR TITLE
[graphql] Add epoch and checkpoint types and some data providers

### PR DESCRIPTION
**Pull request by stefan-mysten**
## Description 

Add epoch and checkpoint types. 

Screenshot 2023-08-18 at 9 48 45 PM

## Test Plan 

Manually for now. 

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-graphql-rpc/tests/snapshots/snapshot_tests__test__schema_sdl_export.snap [hunk 1]
```diff
@@ -93,6 +93,18 @@ type CoinEdge {
...
+type Epoch {
+	epochId: Int!
+	systemStateVersion: BigInt!
+	protocolConfigs: ProtocolConfigs!
+	referenceGasPrice: BigInt!
+	systemParameters: SystemParameters!
+	stakeSubsidy: StakeSubsidy!
+	validatorSet: ValidatorSet!
+	storageFund: StorageFund!
+	safeMode: SafeMode!
```
Comment (oxade) [inline crates/sui-graphql-rpc/tests/snapshots/snapshot_tests__test__schema_sdl_export.snap:105]: Many types added in this struct and others should actually be optional. See the corresponding draft schema entries [here](https://github.com/MystenLabs/sui/blob/main/crates/sui-graphql-rpc/schema/draft_schema.graphql#L395). Not the lack of exclamation mark on the type in the draft schema.
Fixing these should be as easy as changing `field_name: Type` to `field_name: Option<Type>` in the struct definitions, and wrapping the value assignments in `Some(...)`
```
type Epoch {
  epochId: Int!
  systemStateVersion: Int
  protocolConfigs: ProtocolConfigs
  referenceGasPrice: BigInt

  systemParameters: SystemParameters
  stakeSubsidy: StakeSubsidy
  validatorSet: ValidatorSet
  storageFund: StorageFund
  safeMode: SafeMode
```

Comment (stefan-mysten) [reply crates/sui-graphql-rpc/tests/snapshots/snapshot_tests__test__schema_sdl_export.snap:105]: Will fix, thanks!

File: crates/sui-graphql-rpc/src/types/validator.rs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    pub voting_power: Option<u64>,
+    // pub stake_units: Option<u64>,
+    pub gas_price: Option<BigInt>,
+    pub commission_rate: Option<u64>,
+    pub next_epoch_stake: Option<BigInt>,
+    pub next_epoch_gas_price: Option<BigInt>,
+    pub next_epoch_commission_rate: Option<u64>,
+    // pub at_risk: Option<u64>,
+    // pub report_records: Option<Vec<SuiAddress>>,
+    // pub apy: Option<BigDecimal>,
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/validator.rs:40]: APY is not a BigDecimal, it's represented in basis points (an integer between 1 and 10000, measuring 1/100ths of a percent).

I'm actually not sure, how/when it became `Float` in the original schema.  @oxade, do you remember why?  It should be an `Int`.

File: crates/sui-graphql-rpc/src/types/big_decimal.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::str::FromStr;
+
+use async_graphql::*;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) struct BigDecimal(String);
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/big_decimal.rs:9]: As per the comment later on -- we shouldn't need this scalar type.

**Review by amnn — COMMENTED** [review]
My eye was drawn to `BigDecimal` so focussing my comments on that.  As a general comment, this PR is verging on too big, and could be split up into two or three, at least.  it seems silly to put the effort in to splitting up PRs, but time in review scales non-linearly with the length of the PR.

**Comment by stefan-mysten** [comment]
> My eye was drawn to `BigDecimal` so focussing my comments on that. As a general comment, this PR is verging on too big, and could be split up into two or three, at least. it seems silly to put the effort in to splitting up PRs, but time in review scales non-linearly with the length of the PR.

Apologies for another lengthy PR. It was easier to take a stab at once for most of the stuff, than splitting it into small parts.

File: crates/sui-graphql-rpc/src/types/big_decimal.rs (hunk 3 above)
Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/types/big_decimal.rs:9]: That's good to know!

**Comment by oxade** [comment]
@stefan-mysten great job with this PR. 
I'm neutral on it but do you plan on breaking up the PR like @amnn suggested or should we keep reviewing this PR as is?

File: crates/sui-graphql-rpc/src/types/validator.rs (hunk 2 above)
Comment (oxade) [reply crates/sui-graphql-rpc/src/types/validator.rs:40]: > I'm actually not sure, how/when it became Float in the original schema. @oxade, do you remember why? It should be an Int.

I don't quite recall how it came to be a float.

File: crates/sui-graphql-rpc/src/types/epoch.rs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+use super::stake_subsidy::StakeSubsidy;
+use super::storage_fund::StorageFund;
+use super::system_parameters::SystemParameters;
+use super::validator_set::ValidatorSet;
+use async_graphql::*;
+
+#[derive(Clone, Debug, PartialEq, Eq, SimpleObject)]
+pub(crate) struct Epoch {
+    pub(crate) epoch_id: u64,
+    pub(crate) system_state_version: Option<BigInt>,
```
Comment (oxade) [inline crates/sui-graphql-rpc/src/types/epoch.rs:16]: Nit: be consistent on fields being `pub(crate)` vs `pub` for object fields.

**Review by oxade — APPROVED** [review]

File: crates/sui-graphql-rpc/src/types/epoch.rs (hunk 4 above)
Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/types/epoch.rs]: Thanks for spotting this, fixed!

**Pull request merged**